components/supervisor/pkg/supervisor/services.go

```diff
@@ -38,11 +38,42 @@ type RegisterableRESTService interface {
 	RegisterREST(mux *runtime.ServeMux, grpcEndpoint string) error
 }
 
+type ideReadyService struct {
+	ready bool
+	cond  *sync.Cond
+}
+
+func (service *ideReadyService) wait(ctx context.Context) {
+	service.cond.L.Lock()
+	for !service.ready && ctx.Err() != nil {
+		service.cond.Wait()
+	}
+	service.cond.L.Unlock()
+}
+
+func (service *ideReadyService) get() bool {
+	service.cond.L.Lock()
+	ready := service.ready
+	service.cond.L.Unlock()
+	return ready
+}
+
+func (service *ideReadyService) set(ready bool) {
+	service.cond.L.Lock()
```

Comment:
better:
```
service.cond.L.Lock()
defer service.cond.L.Unlock()
```

You can [hold the lock while sending broadcast](https://golang.org/pkg/sync/#Cond.Broadcast).